File: baekjoon/dynamic/2494/main.cpp
#include <iostream>
#include <cstdio>
#include <cstring>

using namespace std;
const int oo = 0x3f3f3f3f;

int N;
int cache[10001][10];
int choices[10001][10];
int values[10001][10];
int S[10001], O[10001];

int solve(int n, int k);
void trace(int n, int k);
int main()
{
    memset(cache, -1, sizeof(cache));
    scanf("%d", &N);
    string s,o;
    cin >> s >> o;
    for(int i=0; i<s.size(); i++) {
        S[i] = s[i]-'0';
        O[i] = o[i]-'0';
    }
    int ret = solve(0,0);
    printf("%d\n", ret);
    trace(0,0);
    return 0;
}

void trace(int n, int k)
{
    if(n == N) return ;
    if(values[n][k] == 0) {
        trace(n+1, k);
    } else {
        printf("%d %d\n", n+1, values[n][k]);
        trace(n+1, choices[n][k]);
    }
}

int solve(int n, int k)
{
    if(n == N) return 0;
    
    int& ret = cache[n][k];
    if(ret != -1) return ret;
    
    int cur = (S[n]+k)%10;
    int o = O[n];
    int left = (o-cur+10)%10;
    int right = (cur-o+10)%10;
    int leftRet = solve(n+1, (k+left)%10)+left;
    int rightRet = solve(n+1, k)+right;
    if(leftRet < rightRet) {
        // n 번째를 왼쪽으로 움직임
        choices[n][k] = (k+left)%10;
        values[n][k] = left;
        ret = leftRet;
    } else {
        // n 번째를 오른쪽으로 움직임
        choices[n][k] = k;
        values[n][k] = -right;
        ret = rightRet;
    }
    return ret;
}

File: baekjoon/dynamic/2494/Main.java
import java.util.*;

class Main {
    static int N;
    static int S[], O[];
    static int cache[][], choices[][], values[][];
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        S = new int[10001];
        O = new int[10001];
        cache = new int[10001][10];
        choices = new int[10001][10];
        values = new int[10001][10];
        for(int i=0; i<=10000; i++) {
            Arrays.fill(cache[i], -1);
            Arrays.fill(choices[i], 0);
            Arrays.fill(values[i], 0);
        }
        N = sc.nextInt();
        sc.nextLine();
        String s,o;
        s = sc.nextLine();
        o = sc.nextLine();
        for(int i=0; i<s.length(); i++) {
            S[i] = s.charAt(i)-'0';
            O[i] = o.charAt(i)-'0';
        }
        System.out.println(solve(0,0));
        trace(0,0);
    }
    
    public static int solve(int n, int k) {
        if(n == N) return 0;
        if(cache[n][k] != -1) return cache[n][k];
        int ret;
        int cur = (S[n]+k)%10;
        int o = O[n];
        int left = (o-cur+10)%10;
        int right = (cur-o+10)%10;
        int leftRet = solve(n+1, (k+left)%10)+left;
        int rightRet = solve(n+1, k)+right;
        if(leftRet < rightRet) {
            ret = leftRet;
            choices[n][k] = (k+left)%10;
            values[n][k] = left;
        } else {
            ret = rightRet;
            choices[n][k] = k;
            values[n][k] = -right;
        }
        return cache[n][k] = ret;
    }
    
    public static void trace(int n, int k) {
        if(n == N) return ;
        if(values[n][k] == 0) {
            trace(n+1, k);
        } else {
            System.out.printf("%d %d\n", n+1, values[n][k]);
            trace(n+1, choices[n][k]);
        }
    }
}